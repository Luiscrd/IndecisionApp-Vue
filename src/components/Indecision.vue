<template>
    <img v-if="img" :src="img" alt="bg">
    <div class="bg-dark">
        <div class="container">
            <div class="indecision-container">
                <div class="input-group mb-3">
                    <span class="input-group-text" id="basic-addon3">Hazme una pregunta</span>
                    <input type="text" class="form-control" placeholder="Preguntame..." v-model="question">
                </div>
                <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
                <div v-if="isValidQuestion">
                    <h2>{{ question }}</h2>
                    <h1>{{ answer }}</h1>
                </div>
            </div>
        </div>
    </div>
</template>

<script>
export default {
    name: 'Indecisdion',
    data() {
        return {
            question: null,
            answer: null,
            img: null,
            isValidQuestion: false
        }
    },
    methods: {
        async getAnswer() {
            try {
                this.answer = 'Pensando...'
                const { answer, image } = await fetch('https://yesno.wtf/api').then(resp => resp.json())
                this.answer = answer === 'yes' ? 'Sii!' : 'Noo!'
                this.img = image
            } catch (error) {
                console.log('IndecisionComponent: ', error)
                this.answer = 'No se pudo cargar del API'
                this.img = null
            }
        }
    },
    watch: {
        question(value, oldValue) {
            this.isValidQuestion = false
            this.img = null
            if (!value.includes('?')) return
            this.isValidQuestion = true
            this.getAnswer()
        }
    }
}
</script>

<style scoped>
img {
    height: calc(100% - 160px);
    left: 0px;
    max-height: 100%;
    max-width: 100%;
    position: fixed;
    top: 90px;
    width: 100vw;
}

.bg-dark {
    display: flex;
    align-items: center;
    justify-content: center;
    height: calc(100vh - 160px);
    max-height: 100%;
    max-width: 100%;
    position: relative;
}

.bg-dark {
    background-color: rgba(0, 0, 0, 0.4) !important;
}

.indecision-container {
    position: relative;
    text-align: center;
    z-index: 0;
}


p {
    color: white;
    font-size: 20px;
    margin-top: 0px;
}

h1,
h2 {
    color: white;
}

h2 {
    margin-top: 190px;
}
</style>