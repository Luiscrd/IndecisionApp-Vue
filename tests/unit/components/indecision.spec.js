import { shallowMount } from '@vue/test-utils'
import Indecision from '@/components/Indecision.vue'

describe('Indecision Component', () => {

    let wrapper;

    let getAnswerSpy;

    global.fetch = jest.fn(() => Promise.resolve({
        json: () => Promise.resolve({
            "answer": "yes",
            "forced": false,
            "image": "https://yesno.wtf/assets/yes/10-271c872c91cd72c1e38e72d2f8eda676.gif"
        })
    }))

    beforeEach(() => {

        wrapper = shallowMount(Indecision)

        getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')

        jest.clearAllMocks()

    });

    test('Debe de hacer mach con el snapshot', () => {

        expect(wrapper.html()).toMatchSnapshot()

    })

    test('Debe de existir el elemnto input', () => {

        const input = wrapper.find('input')

        expect(input.exists()).toBeTruthy()

    })

    test('Al escribir en el input no se debe disparar nada', async () => {


        const input = wrapper.find('input')

        await input.setValue('Hola mundo')

        expect(getAnswerSpy).toHaveBeenCalledTimes(0)


    })

    test('Al escribir en el input el simbolo "?" se debe de lanzar el fech', async () => {

        const input = wrapper.find('input')

        await input.setValue('Hola mundo?')

        expect(getAnswerSpy).toHaveBeenCalledTimes(1)


    })

    test('Pruebas en getAswer', async () => {

        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect(img.exists()).toBeTruthy()

        expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/10-271c872c91cd72c1e38e72d2f8eda676.gif')

        expect(wrapper.vm.answer).toBe('Sii!')


    })

    test('Pruebas en getAswer - Fallo en el API', async () => {

        fetch.mockImplementationOnce(() => Promise.reject('API is down'))

        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect(img.exists()).toBeFalsy()

        expect(wrapper.vm.answer).toBe('No se pudo cargar del API')

    })

})