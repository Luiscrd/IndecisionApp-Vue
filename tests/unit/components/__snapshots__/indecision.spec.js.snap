// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component Debe de hacer mach con el snapshot 1`] = `
<!--v-if-->
<div class="bg-dark">
  <div class="container">
    <div class="indecision-container">
      <div class="input-group mb-3"><span class="input-group-text" id="basic-addon3">Hazme una pregunta</span><input type="text" class="form-control" placeholder="Preguntame..."></div>
      <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
      <!--v-if-->
    </div>
  </div>
</div>
`;
